### PR TITLE
ieee802154: Data frame structs/encoding/decoding

### PR DESCRIPTION
**Pull request by bbbert**
This PR separates out the 802.15.4 header encoding/decoding logic from the radio transmission logic, inserting a new layer of abstraction between the 802.15.4 radio and the userspace driver to achieve that. There are three parts to this PR:

1. A generic buffer management strategy for encoding items to mutable buffers and decoding items from buffers, implemented in `capsules::net::stream`. If this proves to be more generally useful and stable, maybe it can be moved elsewhere.
2. A very complete 802.15.4 header encoding and decoding library built on this buffer management strategy, living in `capsules::net::ieee802154`. This implements most of the general MAC frame header format, which encompasses everything except multipurpose frames (which have a differently-sized frame control header). We do not need multipurpose frames for Thread.
3. The MAC device abstraction, which sits atop the radio and exposes a frame-oriented interface to its user (currently the userspace radio driver). From now on, "user" refers to the client of the MAC device. This way, the radio only has to care about transmitting and receiving, and the user only needs to care about what kind of frame it wants to send. One consequence of this is that the handling of the radio's configuration state via callbacks (waiting on transfers until the radio is done with configuration) is captured in the MAC layer. Hence, the userspace driver no longer needs a config callback.

There are some complicated issues that are quite subtle:
1. Buffer ownership.
  - Transmit: Currently, the user provides the buffer that is sent all the way down to the radio. However, we do not want it to be responsible for figuring out how to prepare the frame. Hence, it calls `mac.prepare_data_frame` on the buffer, which fills in the header fields and produces an auxiliary structure containing just enough state to determine where the payload should be placed. This makes it possible for the user to call `frame_info.append_payload` without knowing where in the buffer to place its payload. When calling mac.transmit_frame, ownership of the buffer is passed to the MAC, and then to the radio, and is returned when transmission is complete.
  - Receive: The buffer used for receiving frames is owned by the radio. This is because frames can be received any time. When a frame is received, the user only gets to see the frame as an immutable reference. In the future, the radio might also be changed to expose the frame only as an immutable reference, if we realize that a copy must occur due to the way decryption works asynchronously. The current procedure of having to call `radio.set_receive_buffer` to return the buffer is brittle and prone to error.
  - Both: Previously, the transmit function did not return the buffer upon encountering an error. Hilariously, this means that if the userspace driver ever tries to transmit while the radio is busy, it loses access to its buffer forever. This PR tries to handle errors more carefully, temporarily by returning `(ReturnCode, Option<&'static mut [u8])`. This should really be some sort of new `ReturnCode` enum that wraps a buffer, but the tuple suffices for now.
2. IEs in the 802.15.4 header. This isn't so much of an issue because we currently don't need to use IEs anywhere except for parsing them properly. As per the discussion in the comments, I went with a fixed-size array of IEs that might slice into the parsed buffer to handle undissected IEs. This might change when security is added, because the incoming frame security procedure might interact with how and when IEs can be parsed. (Payload IEs are part of the encrypted payload.)

Transmission and reception works as usual when testing with the userspace radio apps. We might need to re-think how the userspace driver exposes the details of frames, but this PR does not touch that.

### CODE REVIEW

**Comment by alevy** [comment]
I don't think I understand the background for this very well (what is this for precisely? Which part of the 6lowpan stack would use this? What is the next step after using this interface?)

But... The second approach on the surface seems better. Though, why a linked list of HeaderIE instead of a slice? A linked list is going to be ugly to construct on the stack (all called would need to have the same lifetime, which won't be fun until non-lexical lifetimes land in Rust) and I think it would end up creating a bench of needless reference when they could just be packed together in a slice.

**Comment by bbbert** [comment]
Background: current MAC header parsing functionality is not nearly enough to interoperate with openthread. We need a more complete implementation, not to mention a cleaner approach, otherwise the logic between fragmentation and radio transmission will probably be buggy and unmaintainable.

For header IEs, the issue with slices in the header struct is that when composing a header with IEs, you cannot place IEs in the same buffer as the one you intend to encode the header in, because then you'd need both a mutable borrow and a borrow. This forces you to allocate another temporary buffer to store IEs. As it stands we will already need a lot of buffers flying around the fragmentation/reassembly logic, and having one more would be really bad.

Linked lists, as you mention, have that nasty lifetime issue. We can also decouple IEs from the rest of the MAC header, but that seems to be a little clunky, since they are conceptually part of the MAC header.

Any nicer solution would be much appreciated. We'll probably face the same issue with next header extensions in IP/6lowpan packets and with TLVs.

**Comment by brghena** [comment]
I'd expect the easiest solution here to look something like:

```
pub struct MacFrameHeader {
    header_ies: [HeaderIE],
    payload_ies: [PayloadIE],
}

pub struct HeaderIE {
    type: HeaderType,
    value: [u8],
    ...
}

...
```

Where `header_ies` is an array of `HeaderIE` structs, which could have a length of zero. (note: I don't know what fields a Header IE has, so I made it up)

One big question about the viability of this is the lifetime requirements of your `MacFrameHeader` structs. I'm guessing that what will happen is that the lowest layer of the radio driver will have a `'static` buffer of bytes which it will convert into a Header struct, and the Header struct will be passed up through callbacks and interacted with before either copying it to Userland or being done with it. If that's the case, the Header struct can be created on the stack of the lowest layer and passed upwards through callbacks. So maybe it will need a lot of space for arrays of IEs, but it is only temporarily used stack space.

This is the technique we're using for our in progress Signbus implementation (warning, messy code: https://github.com/helena-project/tock/blob/fe0abb3d07d800e0e58c512a462aaa73fbaa2390/capsules/src/signbus/support.rs#L73). We take the `[u8]`, use it to create a Packet on the stack, and then pass that Packet up to our client, knowing that it will either: perform some action based on it, pass it up to another client, or copy the bytes from it into a userland `AppSlice`, allowing the Packet itself to be `Drop`ped.

**Comment by bbbert** [comment]
@brghena Hmm, interesting. Unfortunately the number of IEs is not known until you walk the chain of IE elements. Wouldn't it have to be `&[HeaderIE]` though, since `[T]` is not `Sized`? I'm not sure how that would work with parsing, because the function would presumably need to return a header struct containing a slice borrow with a lifetime that outlives the function?

**Comment by alevy** [comment]
@bbbert:

I think there is may some details/context missing that's making it hard for us to either asses the solution or make recommendations.

It is perfectly possible to store a field that's a slice with unknown size (you're right that it has to be a reference). It sounds, though, like there is some reason you believe it won't be possible to parse this before creating the struct?

If that's the case, I'm doubtful you'll be able to do it with a linked-list and, in any case, we'll want to have _some_ upper limit. For example, if there are 10,000 headers, we're not going to want to be allocating 10s of KBs on the stack (or anywhere for that matter).

If there is an upper limit, you can always allocate the maximum possible size slice (with `Option`s as the elements possibly) and then slice it.

I'm spit-balling a bit here because I don't think I have a clear sense of specifically what issue you're dealing with (maybe it's in the code, in which case, can you please point to specific places?).

**Comment by bbbert** [comment]
@alevy Sorry, you're right. Specifically it has to do with how these header structs need to be interacted with in the end. My goal is to allow them to be easily consumed after decoding and constructed for encoding. Ideally we want to be able to do something like
```rust
let header = Header { /* ... */ };
header.encode(&mut some_buf);
let header = Header::decode(&buf);
/* easily read header.header_ies */
```

On top of that restriction, we also need to be able to easily obtain the offsets of the beginning of the payload IEs and the end of the payload IEs for the purposes of the encryption process.

So yes, the first solution I mooted was to simply store a byte slice reference, and do the IE parsing somewhere else. However, the downside to that is that creating a header now requires allocating `[u8]` on the stack, like so:
```rust
let some_slice = [0u8; SOME_MAX_SIZE];
let bytes = some_header_ie.encode(&mut some_slice);
let header = Header { /* ... */ header_ies: Some(&some_slice[..bytes]) /* ... */ };
header.encode(&mut buf);
```

That's fine, and is actually the approach I'm pursuing now because we don't actually need to be able to create IEs for the purposes of Thread, just parse them and know where the boundaries are. So just handling two extra offsets would pose no trouble. However, in general, I think it's necessary to find an idiomatic way to deal with headers with arbitrary numbers of repeated units, because this will come up again and again: IPv6 extension headers, Thread TLVs, etc.

The slice-of-structs approach above is another option. However, I may be misunderstanding some Rust concepts, but I don't understand how it would work when decoding.
```rust
// encoding is nice
let header_ies = [HeaderIE { /* ... */ }, HeaderIE { /* ... */ }];
let header = Header { header_ies: &header_ies };
header.encode(&mut buf);
// decoding is not
pub fn decode(buf: &[u8]) -> Header<'what_is_this_lifetime> {
  // look through the chain once (number of IEs is implicit)
  let (ie_start, ie_end, num_ies) = figure_out_where_ies_are(buf);
  let mut header_ies = [HeaderIE; num_ies];
  // look through the chain again
  reparse_and_construct_chain_of_ies(&buf[ie_start..ie_end], &mut header_ies);
  /* ... */
  Header {
    header_ies: &header_ies, // Now what? This lifetime doesn't make sense!
  }
}
```

That approach would need a static or passed-in array of HeaderIEs, I think. That's kind of ugly. I'm not sure what the LinkedList approach would look like, but I presume it would pretty much have the same problem.

The root of the issue is the impossibility of allocating a variable-sized collection without dynamic memory allocation. I hope I'm wrong and there's a better way, but maybe the `&[u8]` approach is actually the only one that works well in malloc-less Tock?

**Comment by alevy** [comment]
OK, that's a really useful explanation. I think I understand a bit better now.

> The root of the issue is the impossibility of allocating a variable-sized collection without dynamic memory allocation. I hope I'm wrong and there's a better way, but maybe the &[u8] approach is actually the only one that works well in malloc-less Tock?

These seem like two orthogonal issues. It's absolutely correct that you cannot allocate a dynamically sized value without a heap (in Rust anyway). But it doesn't matter if the slice has elements of `u8` or of `HeaderIE`. You can allocate a `[HeaderIE; 16]` on the stack just as easily as you can allocate a `[u8; 16]` on the stack (or wherever, statically too).

And I said, we're never going to allow decoding of an completely arbitrary sized header. At an extreme, I don't see how we will ever support parsing a header that has 1000s of IEs because that would just require too much memory to fit.

So there is _some_ reasonable limit. Let's say it's 10.

One reasonable approach is to have the `Header` look like this:

```rust
pub struct Header {
    ...
    headers_ies: [HeaderIE; 10],
    headers_ies_len: usize,
    ...
}

impl struct Header {
    pub fn header_ies(&self) -> &[HeaderIE] {
        self.headers_ies[..self.headers_ies_len]
    }

    pub fn decode(buf: &[u8]) -> Header {
        let mut header_ies = [HeaderIE; 10];
        let header_ies_len = parse_and_construct_chain_of_ies(&buf, &mut header_ies);
        /* ... */
        Header {
            header_ies: header_ies,
            header_ies_len: header_ies_len,
        }
}
```

**Comment by bbbert** [comment]
> These seem like two orthogonal issues.

Oops, one additional thing that maybe wasn't clear: The advantage of `&[u8]` over `&[HeaderIE]` is that you can slice into the input struct, like so:
```rust
struct Header<'a> { header_ies: &'a [u8] }
fn decode<'a>(buf: &'a [u8]) -> Header<'a> { Header { header_ies: &buf[..4] } }
```

The fixed-size limit seems like a viable option. Just to make it clear, there will never be 1000s of IE headers in any case (802.15.4 MTU is 127 bytes, except in certain special applications, non-Jumbogram IPv6 is 1280 bytes). If that's the case, we could even compromise with `enum IEList<'a> { Unparsed(&'a [u8]), Parsed([HeaderIE; 10]) }` with conversion methods `::parse` and `::into_slice(&slice)`.

Note, each `HeaderIE<'a>` will still be an enum with the unparsed u8 slice option for space-efficient undissected IEs. Having a fixed size array for the content field (max 127 bytes) of those IEs would make the HeaderIE struct size unnecessarily large.

**Comment by alevy** [comment]
> Note, each HeaderIE<'a> will still be an enum with the unparsed u8 slice option for space-efficient undissected IEs.

OK. I am 96% lost again.

It sounds like you're saying that there you'd rather just keep the HeaderIEs as bytes slices and parse them later. Your original post made it seem like you _didn't_ want to do that, or maybe the issue you saw with it doesn't seem so problematic any more. In any case it seems like a reasonable approach to me.

**Comment by bbbert** [comment]
@alevy My bad. Let me try to clarify what I meant. 

**tl;dr** The question of whether each HeaderIE can be a byte slice is independent from the question of whether the entire list of IEs is represented as a byte slice.

The core issue is still how to represent variable-length lists of header extensions. To this end, there are two viable options:
- `&[u8]` Borrow a byte slice. Decoding here is good because it initially takes up no extra memory as we can slice into the given buffer. We can then optionally decide to further dissect the byte slice if needed. Encoding here is bad because we need to allocate an additional byte array.
- `[HeaderIE; N]` Fixed-size struct array. Encoding here is good, because we can just fill in the array. Decoding here is good. It would be better if we can limit the amount of extra memory taken up by the fixed-size struct array, and one way to do that is to limit the size of the HeaderIE struct.

There's a really long list of IEs, each with a different format dictated by the element type. We want to dissect those that we need to know, but leave the rest unparsed in their general form: length + type + byte contents. The nicest way to do this is to represent each IE as an `enum` of some known header types and a general unknown or undissected type containing raw bytes. My comment is that this is better achieved with a borrowed byte slice into the buffer instead of a fixed-size byte array. This is because the maximum content field size is 127 bytes. Imagine representing a header with > 10 * 127 bytes when the MTU is only 127 bytes.

Allowing the HeaderIE to be just a borrowed byte slice in those cases is a good way to limit the size of the HeaderIE struct. The complication is that it will then need a lifetime. Hence, using a fixed-size struct array for the IE list does not eliminate the need for a lifetime parameter on the Header struct. I wasn't sure if this would affect any potential solution for the core question, so I brought it up just in case.

I'm currently in favour of the array of structs: a small maximum number of IEs seems to be a valid assumption. I think this is the best approach. With the slicing idea above, it also doesn't take up much extra memory.

I hope that clears things up!

**Comment by bbbert** [comment]
@alevy I adopted the fixed-size array of structs approach, and it seems to work great. I also wrote the MAC layer so that it becomes clear exactly how the header parsing facility will be used, and why I needed to be careful with the lifetimes of the buffers involved. It's in this PR because the MAC layer and the requirements for data frame parsing are very tightly linked, I think.

I think this approach of separating out MAC frame handling from the radio driver makes it much cleaner and easier to implement encryption. I've left out actually calling encryption (we don't have aes-ccm + ccm* yet) but I've structured the code in a way that makes it clear how the asynchronous path will work.

To summarize what each layer is now responsible for:
capsules::rf233::RF233 (kernel::hil::radio::Radio): radio transceiver, automatic ACK, CSMA-CA, CRC generation and validation
capsules::mac::MacDevice (capsules::mac::Mac): frame preparation/transmit/receive, security (in the future), channel scanning (in the future)
capsules::radio::RadioDriver (uses the Mac): shuttling frames to/from userspace

I've also changed how frames are sent back to userspace: as per @phil-levis's suggestion, userspace should receive the whole frame, otherwise it would be impossible to query the length/addresses/pan_ids of the received packets. I'm still not sure what the best way to expose header inspection to userspace is (do we do it in userspace or via calls to kernelspace header parsing?) but that's out of the scope for this PR, I think.

What do you think of the changes I've made to the radio HIL? I'm not sure if the capsules module is the right place to put the trait, but I think it is, assuming a MAC device is a software concept.

**Comment by alevy** [comment]
@bbbert it'll take some time to review these, but I'll try to take a pass ASAP.

One comment off the bat. While we're here, can we rename the `radio` HIL & capsules to something more appropriate? Like `ieee802154` and `ieee802154mac` or something? It's just poorly named right now because 802.15.4 is just one of three radios that Tock supports in some for or another, and there is no relationship between their code or interface.

> I'm not sure if the capsules module is the right place to put the trait, but I think it is, assuming a MAC device is a software concept.

Yes I agree, i think it should be in the same module.

> I've also changed how frames are sent back to userspace

That's fine. I'm skeptical the final userspace API will look anything like what it looks like right now, and as you said, that discussion is for another PR. Whatever is most expedient at the moment.

> I think this approach of separating out MAC frame handling from the radio driver makes it much cleaner and easier to implement encryption. I've left out actually calling encryption (we don't have aes-ccm + ccm* yet) but I've structured the code in a way that makes it clear how the asynchronous path will work.

This all seems great.

**Comment by bbbert** [comment]
> it'll take some time to review these, but I'll try to take a pass ASAP.

Thanks! It's a long one, and probably has bugs.

> One comment off the bat. While we're here, can we rename the radio HIL & capsules to something more appropriate? Like ieee802154 and ieee802154mac or something? It's just poorly named right now because 802.15.4 is just one of three radios that Tock supports in some for or another, and there is no relationship between their code or interface.

Sounds about right. Can we think a little bit about where other things like the header parsing facility should go too? I'd be slightly confused to see both `capsules::ieee802154` and `capsules::net::ieee802154`. Presumably there will be `capsules::net::{lowpan,ip6,udp,coap}` and `capsules::thread::coap` and so on too. Maybe we should have `ieee802154driver` and `ieee802154mac` and `rf233` or something to that effect? I'm not sure there's a good way around this without getting too verbose.

File: capsules/src/mac.rs [hunk 1]
```diff
@@ -0,0 +1,365 @@
+//! Implements 802.15.4 MAC device functionality as an abstraction layer over
```
Comment (alevy) [inline capsules/src/mac.rs]: This comment should be more objective, and more independent of the general design.

It should basically answer the following questions:

1. What functionality does this capsule implement? (ieee 802.15.4 MAC layer, with a reference to the spec as well as specifying any choices this implementation uses that are as general as the spec).

2.  How would I use it? Specifically, how do the exposed structs and traits relate to each other? How (more or less) would a user set this up to use it? What dependencies does is have (i.e. the `Radio` trait, and I think this is the only place to mention the raw radio transceiver)?

File: capsules/src/mac.rs [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+//! TODO: Channel scanning
+
+use core::cell::Cell;
+use kernel::ReturnCode;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil::radio;
+use net::ieee802154::*;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct FrameInfo {
```
Comment (alevy) [inline capsules/src/mac.rs]: Each of the exposed structs should be documented. If the purpose of a particular `struct`, like this one, is covered better in the module comment, just a short summary is fine.

Comment (alevy) [inline capsules/src/mac.rs]: Actually in this case, some detailed explanation in the documentation would be useful since it seems to be a bit of a non-intuitive interface. That detail should likely come in the method comments, but some reasoning in the struct documentation can help set the scene.

File: capsules/src/mac.rs [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    // These offsets are relative to buf[radio::PSDU_OFFSET..] so that
+    // the mac frame length is data_offset + data_len
+    mac_payload_offset: usize,
+    data_offset: usize,
+    data_len: usize,
+    security_params: Option<(Security, [u8; 16])>,
+}
+
+impl FrameInfo {
+    pub fn frame_length(&self) -> usize {
```
Comment (alevy) [inline capsules/src/mac.rs]: Document

File: capsules/src/mac.rs [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        self.data_offset + self.data_len
+    }
+
+    pub fn remaining_data_capacity(&self, buf: &[u8]) -> usize {
```
Comment (alevy) [inline capsules/src/mac.rs]: Document

File: capsules/src/mac.rs [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        buf.len() - radio::PSDU_OFFSET - radio::MFR_SIZE - self.data_offset - self.data_len
+    }
+
+    pub fn append_payload(&mut self, buf: &mut [u8], payload: &[u8]) -> ReturnCode {
```
Comment (alevy) [inline capsules/src/mac.rs]: OK, this the central method of `FrameInfo`, so I'll ask here.


Why not internalize the `buf` into `FrameInfo`, i.e. when a `FrameInfo` is created, it stores the buffer as a field. Then, `append_payload` would just look like this:

```rust
pub fn append_payload(&mut self, payload: &[u8]);
```

I think this would solve a couple issues with the current design:

  1. Currently, the user should make sure to call `prepare_data_frame` and `transmit` with the same `buf`, but nothing is enforcing that.

  2. Currently, I believe, the user could modify the header that `prepare_header` serializes into the buffer before calling `transmit`. That's _maybe_ OK from a safety perspective (I haven't thought about whether anything bad can happen), but it seems simpler if we can just assume the user hasn't touched the header part of `buf` in between calls.

If you do this, it probably makes sense to rename this just to `Frame` or something.

It's also pretty possible I've got this wrong. _But_ if you keep the interface as it is, some documentation somewhere should probably help someone (like me!) understand why.

File: capsules/src/mac.rs [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+        }
+        let begin = radio::PSDU_OFFSET + self.data_offset + self.data_len;
+        buf[begin..begin + payload.len()].copy_from_slice(payload);
+        self.data_len += payload.len();
+
+        ReturnCode::SUCCESS
+    }
+}
+
+pub trait Mac {
```
Comment (alevy) [inline capsules/src/mac.rs]: Document. But also...

Why does this need to be a trait? Is the idea that you may have several different MAC implementations for 802.15.4? If so, it seems better for this to be in `radio.rs` (for now), since it'd be odd to implement a different MAC that depends on this MAC just for the trait.

It's probably also OK to get rid of this for now.

One possible utility for it is if you wanted to mock the MAC layer in order to test higher layers.

Comment (alevy) [inline capsules/src/mac.rs]: btw, document not just the trait itself, but also each of the methods. Most of them are relatively self-explanatory, in which case just a basic comment is fine. But some are less so. For example, what are the units returned from `get_tx_power`? What is the valid range of `get_channel` and `set_channel` and what are the units (i.e. probably channel numbers that map to a frequency)?

There seems to be a relationship between the `set_*` methods and the `config_commit` methods. What is it?

File: capsules/src/mac.rs [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+                          src_addr: MacAddress,
+                          security_needed: Option<(SecurityLevel, KeyId)>)
+                          -> Result<FrameInfo, ()>;
+    fn transmit(&self,
+                buf: &'static mut [u8],
+                frame_info: FrameInfo)
+                -> (ReturnCode, Option<&'static mut [u8]>);
+}
+
+pub trait TxClient {
```
Comment (alevy) [inline capsules/src/mac.rs]: Document.

File: capsules/src/mac.rs [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+    fn send_done(&self, spi_buf: &'static mut [u8], acked: bool, result: ReturnCode);
+}
+
+pub trait RxClient {
```
Comment (alevy) [inline capsules/src/mac.rs]: Document.

File: capsules/src/mac.rs [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum TxState {
+    Idle,
+    Encrypting,
+    ReadyToTransmit,
+}
+
+pub struct MacDevice<'a, R: radio::Radio + 'a> {
```
Comment (alevy) [inline capsules/src/mac.rs]: Document. I suspect this will mostly overlap with the module documentation, so a reference from here or to here is probably fine for the details.

File: capsules/src/mac.rs [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+    fn set_address(&self, addr: u16);
+    fn set_address_long(&self, addr: [u8; 8]);
+    fn set_pan(&self, id: u16);
+    fn set_channel(&self, chan: u8) -> ReturnCode;
+    fn set_tx_power(&self, power: i8) -> ReturnCode;
+
+    fn config_commit(&self) -> ReturnCode;
+
+    fn is_on(&self) -> bool;
+    fn prepare_data_frame(&self,
```
Comment (alevy) [inline capsules/src/mac.rs]: See comment in `FrameInfo#append_payload"

File: capsules/src/mac.rs [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+
+    pub fn set_transmit_client(&self, client: &'static TxClient) {
+        self.tx_client.set(Some(client));
+    }
+
+    pub fn set_receive_client(&self, client: &'static RxClient) {
+        self.rx_client.set(Some(client));
+    }
+
+    fn lookup_key(&self, level: SecurityLevel, key_id: KeyId) -> Option<(Security, [u8; 16])> {
```
Comment (alevy) [inline capsules/src/mac.rs]: I believe this is still todo, right? i.e. it works for a certain static scenario but it's a kind of useless function as it is right now, that's basically here as a stub. If so, a comment would be nice saying that.

File: capsules/src/mac.rs [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+        };
+        Some((security, [0; 16]))
+    }
+
+    fn step_transmit_state(&self) -> (ReturnCode, Option<&'static mut [u8]>) {
+        match self.tx_state.get() {
+            TxState::Idle => (ReturnCode::SUCCESS, None),
+            TxState::Encrypting => {
+                // let frame_info = self.tx_info.get().unwrap();
+                let buf = self.tx_buf.take().unwrap();
```
Comment (alevy) [inline capsules/src/mac.rs]: `unwrap` is a no-no. you can avoid it by doing a `map_or` and returning an error in the `or` case. Or, if it feels better:

```rust
if let Some(buf) = self.tx_buf.take() {
    self.tx_state.set(TxState::Idle);
    (ReturnCode::ENOSUPPORT, Some(buf)
} else {
    (ReturnCode::FAIL, None)
}
```

File: capsules/src/mac.rs [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+                // TODO: implement encryption
+                self.tx_state.set(TxState::Idle);
+                (ReturnCode::ENOSUPPORT, Some(buf))
+            }
+            TxState::ReadyToTransmit => {
+                if self.config_in_progress.get() {
+                    // We will continue when the configuration is done.
+                    (ReturnCode::SUCCESS, None)
+                } else {
+                    let frame_info = self.tx_info.get().unwrap();
```
Comment (alevy) [inline capsules/src/mac.rs]: Same as above w.r.t `unwrap`

File: capsules/src/radio.rs [hunk 14]
```diff
@@ -144,42 +139,66 @@ impl<'a, R: radio::Radio> Driver for RadioDriver<'a, R> {
...
-                    rval = self.radio.transmit(addr, kbuf, len as u8, false);
-                    if rval == ReturnCode::SUCCESS {
-                        self.busy.set(true);
+                    if rval != ReturnCode::SUCCESS {
+                        return;
+                    }
+
+                    // Try to transmit the frame, otherwise at least get the
+                    // frame back.
+                    let res = self.mac.transmit(self.kernel_tx.take().unwrap(),
```
Comment (alevy) [inline capsules/src/radio.rs]: `unwrap` is a no-no. I think it can be totally avoided by putting these calls in the close above that, anyway, maps over `kernel_tx`.

**Comment by alevy** [comment]
@bbbert the new PR comment was extremely helpful, thanks! And the PR is in a really good state, my comments notwithstanding. I'm really excited for this!

File: capsules/src/mac.rs (hunk 5 above)
Comment (bbbert) [inline capsules/src/mac.rs]: Yep, the old reason for this is now obsolete. Initially, I used a `Frame` too, but then decided to pull out the `FrameInfo` struct to keep lifetimes simple when trying to avoid copies when passing around the buffer in-place for cryptography. Now that I know that a copy will always have to occur for crypto, I've reworked the logic to again use `Frame`.

File: capsules/src/mac.rs (hunk 6 above)
Comment (bbbert) [inline capsules/src/mac.rs]: I've documented the `get_*` and `set_*` methods. The idea is that after you call one or more `set_*` functions, you must always call `config_commit` once, otherwise there is no guarantee that the RF233 registers match the MAC software layer's addressing information. I've also made it so that calling `config_commit` when a configuration process is still ongoing will effectively "schedule" another config. This is because the underlying RF233 driver treats a configuration as an atomic operation that cannot be interrupted/restarted, so if new addressing parameters are provided, the configuration process must be repeated.

File: capsules/src/mac.rs (hunk 12 above)
Comment (bbbert) [inline capsules/src/mac.rs]: The idea is that actually the presence of the buffer is coupled with the transmit state. Hence, if that is not true, then it is actually a logic bug and there should perhaps have been an explicit `panic!()`. However, I found a way to express this in terms of an enum, so `TakeCell`s are avoided completely and this logic is expressed as part of the type system instead.

**Comment by bbbert** [comment]
@alevy Sweet, it passes the checks!

Thanks for your really long and detailed review. I've worked through it and documented the MAC layer as detailed as I think it should be (as detailed as what I would have liked some of the legacy Tock code to have been). When security is implemented, much of this will actually change as new states and processes are introduced.

As for whether or not `Mac` should be a trait, I think the conversation a few weeks ago went like this: the MAC device should be a layer on top of the radio so that we can separate frame processing from physical/link layer logic. However, there are more integrated radios that might accomplish more of this logic inline for performance reasons. `MacDevice` implementation might be specific to chips with similar capabilities to the RF233. Hence, `Mac` should be a trait.

In which case, maybe we should re-evaluate which module the trait should be placed in and also rename `Radio`, `MacDevice`, `RadioDriver` et al. in a separate PR.

I also took a look at the `FrameInfo` and `TakeCell` `unwrap`-ing issue and decided to solve them both by switching to a `TxState` that holds more information. This way, there is no need to `unwrap_or_else()` to obtain the state-associated information: it is just immediately bound when `match`ing on the state. I think it makes the stateful logic more explicit, which will be very helpful as the MAC layer gets more complicated. Please take a look.

File: capsules/src/radio.rs [hunk 15]
```diff
@@ -188,35 +203,27 @@ impl<'a, R: radio::Radio> radio::TxClient for RadioDriver<'a, R> {
...
-                    let d = &mut dest.as_mut();
-                    for (i, c) in buf[offset..offset + len].iter().enumerate() {
-                        d[i] = *c;
-                    }
-                    app.rx_callback
-                        .take()
-                        .map(|mut cb| { cb.schedule(usize::from(result), len, 0); });
-                }
-                self.radio.set_receive_buffer(buf);
+            self.app.map(move |app| if app.app_read.is_some() {
```
Comment (alevy) [inline capsules/src/radio.rs]: If we did:

```rust
if let Some(dest) = app.app_read.as_mut() {
    ...
}
```

It would make me sleep much easier, even though I believe in practice they are the same.

File: capsules/src/mac.rs [hunk 16]
```diff
@@ -0,0 +1,566 @@
...
+/// functionality:
+///
+/// - Configuration of addresses and transmit power
+/// - Preparing frames (data frame, command frames, beacon frames)
+/// - Transmitting and receiving frames
+///
+/// Outlining this in a trait allows other implementations of MAC devices that
+/// divide the responsibilities of software and hardware differently. For
+/// example, a radio chip might be able to completely inline the frame security
+/// procedure in hardware, as opposed to requiring a software implementation.
```
Comment (alevy) [inline capsules/src/mac.rs]: Got it. OK this makes sense.

File: capsules/src/mac.rs [hunk 17]
```diff
@@ -0,0 +1,566 @@
...
+    /// channel `11 <= channel <= 26`, otherwise EINVAL will be returned
+    fn set_channel(&self, chan: u8) -> ReturnCode;
+    /// Set the transmission power of the MAC device, in dBm. `power` should
+    /// satisfy `-17 <= power <= 4`, otherwise EINVAL will be returned
+    fn set_tx_power(&self, power: i8) -> ReturnCode;
+
+    /// This method must be called after one or more calls to `set_*`. If
+    /// `set_*` is called without calling `config_commit`, there is no guarantee
+    /// that the underlying hardware configuration (addresses, pan ID) is in
+    /// line with this MAC device implementation.
```
Comment (alevy) [inline capsules/src/mac.rs]: Got it, so a legal instance of the `Mac` trait could implement `config_commit` as a noop if all the `set_*` methods committed each change.

**Comment by alevy** [comment]
Ready? Set? Merge...

**Comment by niklasad1** [comment]
@bbbert 

Do you think you can fix the following compiler warnings (that I think this PR has introduced)?
```
warning: variant is never used: `Encrypting`
   --> /home/niklasad1/Projects/tock/capsules/src/mac.rs:247:5
    |
247 |     Encrypting(FrameInfo),
    |     ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: #[warn(dead_code)] on by default

warning: unused variable: `info`
   --> /home/niklasad1/Projects/tock/capsules/src/mac.rs:321:45
    |
321 |                     TxState::ReadyToEncrypt(info, buf) => {
    |                                             ^^^^
    |
    = note: #[warn(unused_variables)] on by default
```

**Comment by alevy** [comment]
FYI, this second one is fixable by prefixing `info` as `_info` (which just tells the compiler that you know it's not used). The first one is more annoying, because it's technically used, but never assigned. Adding a `#[allow(dead_code)]` right above it will do it. We'll just have to remember to remove it when we actually do use it.

Sending a separate PR.

**Comment by ppannuto** [comment]
> We'll just have to remember to remove it when we actually do use it.

Better to just comment it out until it's used? Or is there a complimentary
warning for unnecessary use of allowing dead_code?
